PLT-4176: Implementation of Pinned Posts

Zhouzi opened the pull request:
#### Summary

This PR implements the "pinned posts" feature according to the [spec](https://docs.google.com/document/d/1U29VQGEFEI7JDaTBjvk5xm59MvJD-JMI4TQv1erjnrU/edit#). It allows users to pin a post that is then accessible in the RHS thanks to a button in the channel's header.

#### Ticket Link
- [PLT-4176 - GitHub](https://github.com/mattermost/platform/issues/4035)

#### Checklist
- [x] Add `Pin in <channel>` button
- [x] Create `/pin` web service
- [x] Emit websocket event to refresh pinned posts
- [x] Add unit tests
- [x] Add documentation
- [x] Add `Unpin in <channel>` button
- [x] Create `/unpin` web service
- [x] Emit websocket event to refresh pinned posts
- [x] Add unit tests
- [x] Add documentation
- [x] Create `/pinned` web service to retrieve pinned posts
- [x] Add unit tests
- [x] Add documentation
- [x] Add "pinned posts" button in the channel header
- [x] Add "pinned posts" RHS
- [x] Refresh "pinned posts" RHS when pinning/unpinning a post
- [x] Add `Pin in <channel>` button in the RhsRootPost and RhsComment
- [x] Add `Unpin in <channel>` button in the RhsRootPost and RhsComment

The PR documenting the different endpoints is available here: [mattermost-api-reference/pr/62](https://github.com/mattermost/mattermost-api-reference/pull/62/files)

#### Review Checklist

* [x] Text in the [...] drop-down menu should say `Pin to channel` instead of `Pin in [CHANNEL_NAME]` and `Un-pin from channel` instead of `Un-pin in [CHANNEL_NAME]`
* [x] Update pin icon according to the spec
* [x] Update pinned posts RHS title
* [x] Update pinned posts RHS explanation
* [x] Add "view pinned posts" in mobile nav
* [x] The pin icon in the header moves when clicked
* [x] Add pin/un-pin buttons in the permalink view
* [x] Add a "pinned" indicator in the pinned post
* [x] Properly update the post's pinned status in permalink view
* [x] Fix icons whitespaces in channel header
* [x] Fix the "Pinned" badge spacing in compact view
* [x] Vertically align the "Pinned" badge relative to the timestamp
* [x] Fix the "Pinned" badge position in successive replies

mattermod commented:
Thanks @Zhouzi for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

jasonblais commented:
Thanks @Zhouzi, wonderful work! Really excited about this pull request!

I noticed you have a few items unchecked, is this PR a work in progress? if so, that totally works, I can add a 'Work In Progress' label so others know.

Zhouzi commented:
Thanks @jasonblais ;) Yeah, it's still a work in progress. Should I ping you when it's ready for review? Or maybe just send a message in the "contributors" channel?

jasonblais commented:
Thanks @Zhouzi! You can message here or ping me in the "contributors" channel, either way works.

Really excited!! Huge thanks for working on this feature!

jasonblais commented:
Hey @Zhouzi, thanks so much for working on the pull request, love the progress you've had past week!

Just wanted to give you a heads up about upcoming dates for our 3.5 release:
 - Wed 10/26: last date to submit a pull request for major features
 - Mon 10/31: last date any major features will be merged for 3.5 release [assuming the pull request was submitted by the 26th]

If you'd like to work on getting the pinned posts feature in for 3.5, those would be the dates to meet -- just wanted to let you know so it doesn't come as a surprise if you were aiming for the 3.5 release.

Once again, big thanks for your wonderful work, really excited for this feature!

Zhouzi commented:
Hey @jasonblais, thanks for keeping me updated. I am afraid the PR won't be ready for Wed 10/26 but I'll speed up in November ;)

jasonblais commented:
Sounds great, thanks @Zhouzi! Really excited for this feature, looking forward to it in November!

jasonblais commented:
Hey @Zhouzi, just checking if you need any help or further information?

Thanks again for working on this improvement!

Zhouzi commented:
Hey @jasonblais! Sorry for the delay, was on holidays but I'm now back and plan on finishing the last details of this feature soon.

In fact I am struggling with the attached post as I am not sure how I should be fetching it. I'll think again about the different possibilities and eventually come back to you for some help.

jasonblais commented:
Thanks @Zhouzi, not to worry!

Would be happy to help with any questions you might have. You can post them here or in our Mattermost Contributors channel](https://pre-release.mattermost.com/core/channels/tickets).

jasonblais commented:
Hey @Zhouzi, 

> In fact I am struggling with the attached post as I am not sure how I should be fetching it. I'll think again about the different possibilities and eventually come back to you for some help.

Let me know if you've got a few possibilities and if we can help with brainstorming different options?

Once again, thanks for all your help thus far!

Zhouzi commented:
@jasonblais so here's what is currently happening when a user pin a post (in this PR):

1. The pinned post is updated with a `isPinned` value set to `1`
2. A "system post" is generated with a special property `PinnedPostId` with the pinned post's id.

The issue now is that I am struggling to find the right way to fetch the posts. What I am thinking of is, when posts for a channel are requested, look for posts that have a `PinnedPostId`, query it and add it as a `PinnedPost` to the post. Indeed, it also need to add the author to the pinned post object.

But it kind of brake the data normalization as the pinned post may be available at two different places on the client.

What's your thoughts? Would it be okay to do it that way?

Regarding the RHS, I created an api endpoint that returns the list of pinned posts which may raise a few concerns but I guess it's ok for this purpose.

jwilander commented:
Hi @Zhouzi, I can help you with the technical questions.

Let me reiterate to make sure I understand your issue:

1. A user pins a post
2. A system message referring to the pinned post is added to the channel
3. If the pinned post isn't loaded when the system message is, then you have a problem?

If that's the case then there's two possible solutions I would suggest:

1. Change the design so that we don't need the original post for the system message, maybe @jasonblais can chime in on this. I think it would be preferred because it reduces complexity but it's up to you guys
2. In actions on the client, when loading posts check each post for a PinnedPostId, check if we have it locally yet and then fetch any pinned posts we're missing. We do this similarly for fetching user profiles and statuses for posts in `webapp/actions/post_actions.jsx` so you can use those as an example. This should be done on the client for the sole reason that we should not build the server to match the UI.

Let me know if you have any questions about any of that

yangchen1 commented:
@Zhouzi @jwilander Yes, I definitely we can change the design. We don't need to actually show the original post - an in-line preview similar to how replies look like is fine (the preview links to the pinned message, like below)

![image](https://cloud.githubusercontent.com/assets/5504605/21159689/1fce764c-c150-11e6-956c-a64c7ded034d.png)

yangchen1 commented:
Hey @Zhouzi, the product team discussed we can skip the system message for now so we can get the main functionality in for 3.6 - we'll have it be implemented later in a separate ticket, hopefully for 3.7. :) Hope that makes things easier!

Zhouzi commented:
@yangchen1 I am so sorry for not keeping you guys updated on the status of this PR and really appreciate the efforts for helping with it.

I've been unable to work on the feature because I am on holidays but no excuses, I will work on the last bits, remove the system message, rebase, update and let you know when it's up for review (this week).

yangchen1 commented:
No worries @Zhouzi! We appreciate your work on this. As usual, let us know if you have any questions or need help!

Zhouzi commented:
@yangchen1 I was hoping to have the feature done this week so I added unit tests, [documentation](https://github.com/mattermost/mattermost-api-reference/pull/62) and was about to clear the last rebase issues due to the changes that happened since I created this PR but it looks like the whole flow changed quite a bit.

I need to get familiar with the new apis before I can make the changes so I'll need more time.

In the mean time, do not hesitate to let me know if you guys feel like someone else could work faster on the feature, I'd understand that 😉 

Otherwise, just let me know when it'd be great to have it ready for review so I can tell you honestly if I'd be able to do it.

P.S.: thanks for being so supportive, the experience of contributing to mattermost is really awesome 🎉

esethna commented:
Thanks @Zhouzi, there is no immediate rush given that we're locking down `master` for the v3.6 release and the Mattermost team will be focused on [release process](https://docs.mattermost.com/process/release-process.html) for the first two weeks of January. Please feel free to work on this when you have time and keep us in the loop on any progress, the product management team is around to help with any design questions. Thanks again for your contribution, we are excited about this feature!

Zhouzi commented:
I am finally back from vacations and will be working on it this week :)

By the way, I was pleasantly surprised to find a Mattermost mug in the mail so thank you a lot guys for that, it's awesome 🎉  One thing though, there's the name of someone else on it 😜 Get in touch at hello@gabinaureche.com if you want to sort things out.

jasonblais commented:
Yay! Thanks @Zhouzi!  Looking forward to testing this out soon! :tada: 

I'm really sorry about the mug though, we'll look into it right away.

Zhouzi commented on api/post.go:
```diff
@@ -183,6 +185,86 @@ func updatePost(c *Context, w http.ResponseWriter, r *http.Request) {
+		c.Err = result.Err
+		return
+	} else {
+		post := result.Data.(*model.PostList).Posts[postId]
```
Should I:

* Get the post again via `app.Srv.Store.Post().Get(postId)` and deal with possible errors
* Take the updated post from `app.Srv.Store.Post().Unpin(postId)`'s Data

Solution 2 looks good but not sure if that's its responsibility.

jwilander replied:
Hi @Zhouzi, what about instead of having specific SQL functions and queries that update the post, you get the post first, modify it, then update it with the existing functionality? That way you will have the post to return and there's less code overall since you don't need to add new SQL functions

Zhouzi replied:
@jwilander that's what I used to do but wasn't sure if it was the right way to :) But it's definitely making things simpler, thank you will do that!

Zhouzi commented on store/sql_post_store.go:
```diff
@@ -156,6 +156,82 @@ func (s SqlPostStore) Update(newPost *model.Post, oldPost *model.Post) StoreChan
+			}
+
+			if result.Err == nil {
+				result.Data = postId
```
Obviously this line throws an error as `Data` is not expecting a string. So, should I call SqlPostStore's Get to retrieve the updated post?

Zhouzi commented on store/sql_post_store.go:
```diff
@@ -156,6 +156,82 @@ func (s SqlPostStore) Update(newPost *model.Post, oldPost *model.Post) StoreChan
+			}
+
+			if result.Err == nil {
+				result.Data = postId
```
Same issue as above ☝️

Zhouzi commented on webapp/sass/layout/_content.scss:
```diff
@@ -23,9 +23,9 @@
+        .header-list__members,
+        .header-list__pinned-posts {
             margin-right: -18px;
-            float: right;
```
I had to remove the float right as the icons wouldn't stay on the same line when resizing down the screen. I played around to see if it had any undesired effects but didn't find anything.

jwilander replied:
@asaadmahmood any idea if this will cause any issues?

asaadmahmood replied:
Not sure, need to set up a spinmint once tests pass so that I can investigate.

Zhouzi commented on webapp/utils/constants.jsx:
```diff
@@ -383,6 +385,10 @@ export const Constants = {
     FLAG_ICON_SVG: "<svg width='12px' height='12px' viewBox='0 0 48 48' version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' xml:space='preserve' style='fill-rule:evenodd;clip-rule:evenodd;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:1.41421;'> <g> <g transform='matrix(1,0,0,0.957537,-0.5,1.42123)'> <path d='M2.5,0.5C1.4,0.5 0.5,1.4 0.5,2.5L0 ...
+
+    // TODO: this svg has some size issues (especially viewBox)
+    PIN_ICON_SVG: "<svg width='12px' height='12px' viewBox='0 0 1792 1792' version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' xml:space='preserve' style='fill-rule:evenodd;clip-rule:evenodd;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:1.41421;'><path d='M800 864v-448q0-14-9-23t-23-9-23 9-9 23v448q0 14 9 23t23 9 23-9 9-23zm672 352q0 26-19 45 ...
```
Where do you guys get the svg icons for font-awesome? I found something that's generating them but there's still an issue with the viewBox.

jwilander replied:
@asaadmahmood can probably help more, but you can use the font awesome icons through CSS classes like this:
```
<div className='status'><i className='fa fa-lock'/></div>
```

asaadmahmood replied:
Yeah, you can just use it like that. We use svgs when icons aren't available, see if you can find corresponding icons here: https://fortawesome.github.io/Font-Awesome/icons/
If not, let me know and I'll create an svg icon for you.

Zhouzi commented:
@jasonblais and here we go, the PR is ready for review 🎉  The build seem to fail but I am not sure whether it is related to this PR or not, the error is unclear. Unit tests are passing and there are no lint errors ^^.

I added two comments, one is a note and the other one a question.

I'll be waiting for the review/comments/requests :)

jasonblais commented:
Yay! 🎉  Thanks @Zhouzi, this is exciting! 

@jwilander, would you know what's causing the build errors? Also, any thoughts on the two questions above?

jwilander commented:
The build error is caused by this PR being out-of-sync with some enterprise changes. A rebase should fix the issue (looks like one is needed anyways)

Zhouzi commented:
@jwilander I rebased the branch and it is now up to date with master. The build is still not passing and I didn't figure out what's failing:

```
...
[2017/01/30 18:38:11 UTC] [INFO] Stopping Server...
[2017/01/30 18:38:11 UTC] [INFO] Closing SqlStore
--- PASS: TestZZWebSocketTearDown (2.00s)
FAIL
[2017/01/30 18:38:11 UTC] [INFO] stopping websocket hub connections
[2017/01/30 18:38:11 UTC] [INFO] Server stopped
coverage: 60.4% of statements
exit status 1
FAIL	github.com/mattermost/platform/api	369.794s
...
```

Would you have any advice on how I can resolve this?

hmhealey commented:
@Zhouzi The console logs on Jenkins are sometimes truncated. Right at the top, there's a "Skipping 361KB, [Full Log](https://build.mattermost.com/job/mattermost-platform-pr/4428/consoleFull)" link that you can click to see everything. There's some unit tests failing for you

jasonblais commented:
Hey @Zhouzi, looks like it needs just one more rebase and we'll be able to test it out

![image](https://cloud.githubusercontent.com/assets/13119842/22628000/23141d56-eb9a-11e6-9de9-6803d1aa2ae7.png)

Zhouzi commented:
Hey @jasonblais, actually I noticed that one thing was missing in the wording ("Pin in Town Square" instead of "Pin in Channel") so I had a little more work to do before the rebase but we're all good now :)

There's one thing to note though: there are situations where a post is viewed outside of the context of its channel (e.g when accessing a permalink). It means the channel's data is not accessible, making it difficult to add the `Pin in {channelDisplayName}` button. So as it stands, you cannot pin a post when viewing it from a permalink (the pin/unpin buttons are not added to the dropdown). I could probably get the channel's data somewhere but I am unsure whether that would make sense from a UX perspective or not (you're not in a channel after all).

Anyway, I'll be waiting feedback to make the requested changes :)

jasonblais commented:
Hooray, thank you @Zhouzi!!

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0a18b1ea3940119d8.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0a18b1ea3940119d8

jasonblais commented:
Thanks @Zhouzi, overall this feature is working great! Huge thanks for your contribution!

Just a few notes that I think [you may have accidentally missed on the spec](https://docs.google.com/document/d/1U29VQGEFEI7JDaTBjvk5xm59MvJD-JMI4TQv1erjnrU/edit#):

1) A system message should indicate that the user pinned a message --> after a user pins a message, the following system message should appear:

![image](https://cloud.githubusercontent.com/assets/13119842/22663881/714c514a-ec7c-11e6-982e-e87ab100fcb4.png)

2) Text in the [...] drop-down menu should say
  - `Pin to channel` instead of `Pin in [CHANNEL_NAME]`
  - `Un-pin from channel` instead of `Un-pin in [CHANNEL_NAME]`

This uses correct grammar and avoids corner cases with long channel names (see sample)

![image](https://cloud.githubusercontent.com/assets/13119842/22669234/b8b6b434-ec91-11e6-9880-07372805e2eb.png)

3) The chosen pin icon was different in the spec. Did you intentionally use a different icon for it?

![image](https://cloud.githubusercontent.com/assets/13119842/22669327/0543030c-ec92-11e6-8a38-7e52946c2887.png)

4) After opening the list pinned posts, the header in the right hand side should say `Pinned posts in [Channel_ui_name]` instead of `Pinned Posts`

![image](https://cloud.githubusercontent.com/assets/13119842/22669380/453140e6-ec92-11e6-839c-10e8ceec5aa2.png)

5) Propose changing the third bullet in pinned posts spec with

```
Pinned messages are accessible by all channel members and are a way to mark messages for future reference.
```

instead of 

```
Pinned messages are accessible by all the members of the channel and are a way to mark messages for future reference.
```

![image](https://cloud.githubusercontent.com/assets/13119842/22669399/5d40f208-ec92-11e6-86b5-d9563f5c93dc.png)


6) On mobile, View Pinned Posts is an option under the channel drop-down: “View Pinned Posts” is placed under “View Info” in the dropdown.

![image](https://cloud.githubusercontent.com/assets/13119842/22669552/1cf6c35c-ec93-11e6-8c64-a10974e1edfb.png)

7) The pinned icon seems to move after clicking on it

![image](https://cloud.githubusercontent.com/assets/13119842/22669700/e734b778-ec93-11e6-8917-2b381755c583.png)

![image](https://cloud.githubusercontent.com/assets/13119842/22669710/ed43b8f8-ec93-11e6-87a8-46d7329fd4dd.png)

Zhouzi commented:
Actually I've been following the spec from [PLT-4176 - GitHub](https://github.com/mattermost/platform/issues/4035), I didn't have access to this google document (I just sent a request for it).

Anyway, thank you for the detailed feedback :) I have a few questions:

> A system message should indicate that the user pinned a message

We had decided not to include this feature for now to avoid further delay ([see comment](https://github.com/mattermost/platform/pull/4217#issuecomment-267201458)). Let me know if you want me to implement it.

> The chosen pin icon was different in the spec

The icon in the spec is not part of the icon set that's currently in use so I choose the closest alternative. Let me know if I should change it ;)

> After opening the list pinned posts, the header in the right hand side should say Pinned posts in [Channel_ui_name]

What should I do regarding the permalink view? Should it still be possible to view the channel's pinned posts when viewing a post via its permalink? (This view is not really tied to a channel).

Other issues are fixed or in progress ;)

jasonblais commented:
Thanks @Zhouzi for the quick reply!

> We had decided not to include this feature for now to avoid further delay (see comment). Let me know if you want me to implement it.

Ah I see, I didn't notice that. That sounds good.

I'll queue an item for UX meeting to discuss the system message. I'm wondering if there is a better way to give feedback for the user.

> The icon in the spec is not part of the icon set that's currently in use so I choose the closest alternative. Let me know if I should change it ;)

Got it, let's keep it for now and I'll check with our UI designer about it later

> What should I do regarding the permalink view? Should it still be possible to view the channel's pinned posts when viewing a post via its permalink? (This view is not really tied to a channel).

I think so, unless it's technically complex? We let the user to view replies, flagged posts, etc. in permalink view, so it would make sense to also be able to view pinned posts too.

Which brings to mind: we probably would want to add the ability to pin/un-pin posts in permalink view.

![image](https://cloud.githubusercontent.com/assets/13119842/22753517/602cc342-ee0a-11e6-8f8c-fb853c14cefb.png)

activedecay commented:
This is the only feature that is "No" on a long, long list of comparisons between chat services. I am eager for this to be available.

jasonblais commented:
Hey @Zhouzi, a quick update for your three questions/notes above:

1. We're investigating a few potential options for giving feedback to the user after a message is pinned to the channel (something other than a system message). The discussion will continue on Monday,

2. and 3. let me know if you have any further questions/thoughts about them.

Zhouzi commented:
Thanks for the update @jasonblais :)

I hope to have the remaining issues fixed today.

Adding pinned posts to the permalink view is a bit complex but I'll figure it out, I believe user experience should not suffer from technical details ;)

jasonblais commented:
Thanks @Zhouzi! We'll work on an updated design for 1) and I'll post here with the final design.

One thing we're considering is adding a "pinned by you" indicator after you pin a message, such as shown below. 

![image](https://cloud.githubusercontent.com/assets/13119842/22865038/e6876092-f129-11e6-8ead-5de56afdbee6.png)

Zhouzi commented:
Okay but if you think the system message is the best indicator just let me know and we should be able to sort things out ;)

jasonblais commented:
Hey @Zhouzi,

Can you help 

1) Add a "Pinned" indicator which appears between the timestamp and the flag icon.

The colours for the indicator should match Account Settings > Display > Theme > Link and button styles: Button bg, and button text

![image](https://cloud.githubusercontent.com/assets/13119842/22895269/d52fa1b8-f1e9-11e6-83fe-e06b2b71ac7e.png)


![image](https://cloud.githubusercontent.com/assets/13119842/22891683/240a13d8-f1de-11e6-8d4d-8ed4e72623b9.png)

2) Can you help update the pinned posts icon to match the icon from the spec?

![image](https://cloud.githubusercontent.com/assets/13119842/22895506/c18c7b9e-f1ea-11e6-8042-30f228664519.png)

If you need an SVG icon for it, let me know.

Did you have the chance to update the PR based on previous notes?

Zhouzi commented:
Hello @jasonblais I edited my first post to include the review checklist, there are only 2 items remaining: the icon and "pinned" label.

I'll need the pin icon's svg please :)

jasonblais commented:
That's a great idea! Nice work, I love checklists :) 

I've pinged @asaadmahmood to share you the icon's svg

asaadmahmood commented:
Updated the icon in the [Jira ticket](https://mattermost.atlassian.net/browse/PLT-4176).

Zhouzi commented:
@jasonblais it's all fixed :)

I also updated the usage description from:

> You can pin a  message by clicking the [icon] from the message's menu.

To:

> You can pin a message by clicking the "Pin to channel" option from the message's menu.

Since there are no icons in the menu.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-03012c7fd9e2f6954.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-03012c7fd9e2f6954

jasonblais commented:
Thanks @Zhouzi! 

### Functionality

Functionally works really well, only found one issue:

1) Search for a post
2) In search results, click "Jump" next to the post
3) In permalink view, click the [...] menu and choose "Pin to channel"

Observed: Nothing appears to happen, although the post was successfully pinned. The [...] menu continues to show "Pin to channel" option.

Expected: "Pinned" indicator appears and [...] menu now shows "Un-pin from channel" option.

### UI

Couple of issues:

1) Can you help make the whitespace between 
 a) Channel member list icon and pinned posts icon; pinned posts icon and search bar
equal the whitespace between 
 b) Search bar and recent mentions icon; recent mentions icon and right edge of the Mattermost window

![image](https://cloud.githubusercontent.com/assets/13119842/23105497/4baa9f32-f6ae-11e6-9134-804fbfa332ae.png)

2) In compact view (set from Account Settings > Display > Message Display > Compact), 
  - extra whitespace between username and pinned indicator
  - not enough whitespace between pinned indicator and message contents

Screenshot below with observed and expected. Note that the whitespace left and right of the pinned indicator should be equal

![image](https://cloud.githubusercontent.com/assets/13119842/23105542/05bc74cc-f6af-11e6-9304-50ee684cb88f.png)

3) The pinned indicator (in either Standard or Compact view) is not centered relative to the timestamp

![image](https://cloud.githubusercontent.com/assets/13119842/23105559/2fa3c9b6-f6af-11e6-8676-c1e3807878af.png)

4) The pinned indicator overlaps the message contents in Standard view if you have successive replies and/or messages from the same user

Observed:
![image](https://cloud.githubusercontent.com/assets/13119842/23105571/5e1c9566-f6af-11e6-979f-43ab8c6b2255.png)

Proposed change: Match the UI with the Compact view where the "Pinned" indicator appears first, followed by the message contents

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b88cd2fb8657a518.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b88cd2fb8657a518

jasonblais commented:
Hey @Zhouzi, let me know if you have any questions on the above.

Would love to ship this feature in the next release! If we do, we'd need it merged by Feb 28th, not sure if that's a reasonable timeline?

Zhouzi commented:
Hey @jasonblais the reported issues should be fixed by Feb 24th (maybe before). If everything's working as expected after that, from what I understood we'll go through a dev review. So unless the next phases involve lots of changes, we may be on time for Feb 28th.

jasonblais commented:
Yay, that would be perfect! Thank you @Zhouzi, appreciate it 👍

jasonblais commented:
@Zhouzi Thanks for the latest changes! Do these cover the cases above?

If so, can you help rebase to resolve merge conflicts, and I'll give this another review?

Zhouzi commented on webapp/stores/post_store.jsx:
```diff
@@ -259,22 +260,32 @@ class PostStoreClass extends EventEmitter {
     }
 
+    // updates/adds all the instance of "post" in every postsInfo
     storePost(post, isNewPost = false) {
```
Just for information:

Until now this method used to update/add the post only in its channel. The problem was that in the permalink view the post is linked to the post being viewed, meaning postsInfo contains not only channel ids but also posts ids. With this PR, storePost will go through all the postsInfo that contains the post being stored and update/add it.

jasonblais replied:
@Zhouzi it looks like I'm having trouble getting any posts through. Wondering if it's related to the changes you described above? I worry there will be performance issues for large deployments, if `storePost` goes through all the `postsInfo` that contains it.

![image](https://cloud.githubusercontent.com/assets/13119842/23324328/0d501d78-fabb-11e6-9d00-89c690ad1d1a.png)

If it's easier, I'd propose instead removing the `Pin to channel` menu item from [...] when viewing a message in permalink view. This is because
 - it will involve fewer changes (and none to database)
 - the use case for pinning messages in permalink view is probably low (if not, we'll get feedback and we can add this functionality later)

Sorry that I'm proposing an alternate change, I wasn't aware of how significant the change would be.

Zhouzi commented:
@jasonblais Here you go, I've fixed the last issues.

I've tweaked the different badges to better align them vertically and hopefully it's better now:

![screen-shot-2017-02-24-at-22 53 34](https://cloud.githubusercontent.com/assets/2291025/23322695/6cc34cd8-fae5-11e6-9e19-c1e96e20450b.png)

To be honest there are quite a few edge cases where the badge has to be slightly moved so I tried to play around with different view modes, window sizes, comments/replies/messages and so on but something may have escaped my attention. So let me know if you spot anything weird :)

jasonblais commented:
This is awesome, thank you @Zhouzi! Will test soon

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0f03af9f3d41ea2aa.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0f03af9f3d41ea2aa

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0718b9dec3bb6cd0a.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0718b9dec3bb6cd0a

Zhouzi commented:
@jasonblais sorry, a last minute change broke the message submission but it's fixed now. I've also changed a bit how the posts are updated: it's only looking into two places (the post's channel and permalink view) to make the appropriate updates, so no performance issues :)

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0c6810a8842ae2abc.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c6810a8842ae2abc

jasonblais commented:
Thanks @Zhouzi, just three notes. Would you be able to help take a look by today or tomorrow and I'll then escalate this to dev review?

There is one UI issue I found in compact mode (in mobile), but I'm not sure how to solve that. I'll ask our UI designer on Monday who can help, so no need to worry about. Just posting here so you're aware.

Really solid work 👍 

1) The "Pinned posts in [channel_name]" breaks if the channel name is long.

![image](https://cloud.githubusercontent.com/assets/13119842/23332985/5a4f83fc-fb51-11e6-8f7d-9cabb73d7e23.png)

Propose if the channel name is too long, we truncate it with "..." similar to how we truncate long usernames on the right-hand side.

![image](https://cloud.githubusercontent.com/assets/13119842/23333005/bc755e1c-fb51-11e6-9068-afdc19c610b6.png)

2) JavaScript errors for "channel_id undefined"

It appears randomly, but I'm able to reproduce it reliable upon login and upon refresh. Couldn't reproduce on `master` branch.

![image](https://cloud.githubusercontent.com/assets/13119842/23333047/79e6a5fa-fb52-11e6-9f11-1d0512bd3d7b.png)

3) Position of timestamp

The screenshot you showed looks good, but on the test server it doesn't look like it's quite centered

![image](https://cloud.githubusercontent.com/assets/13119842/23333101/b9806790-fb53-11e6-9e2f-1e4f8b3b0759.png)

Zhouzi commented:
@jasonblais could you please tell me what browser you are using? I am unable to reproduce the badge's vertical alignment issue.

I've fixed the other issues but while doing so I've noticed two other bugs:

The inline "Pinned" badge is still overlapping the text in Safari:

<img width="174" alt="screen shot 2017-02-25 at 22 35 41" src="https://cloud.githubusercontent.com/assets/2291025/23334969/c8852724-fbaa-11e6-8ae5-4d9da6da43cd.png">

The channel header's actions are moving over the RHS header in certain screen size:

<img width="398" alt="screen shot 2017-02-25 at 22 35 14" src="https://cloud.githubusercontent.com/assets/2291025/23334973/d67c3610-fbaa-11e6-999d-8396f833697a.png">

Those bugs are likely to require refactoring some existing CSS code and thus further testing. To be honest, I'm starting to doubt whether the PR is going to be on time or not. I'll continue working on it tomorrow but don't know if I'll be able to have it done without breaking anything.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0d29e7aa8762b408d.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0d29e7aa8762b408d

jasonblais commented:
@Zhouzi Thank you for the quick reply, I verified both issues 1 and 2 are fixed.

For the CSS fixes, let's try to work to where we can tomorrow, and then I'll bring our UI/CSS expert to help with the rest on Monday. How would that sound like? 

Functionally it's working great, and your work has been fantastic - I wouldn't want to block the release of this awesome feature since it's so close :) 

PS: The one other case I saw was the "pinned" indicator overlapping the "..." menu on mobile view

![image](https://cloud.githubusercontent.com/assets/13119842/23335033/021fb53e-fb7a-11e6-9594-c1fe55143ebb.png)

jasonblais commented:
^oh, and for the overlap: I'm testing on Windows 10, Chrome

Zhouzi commented:
@jasonblais I managed to fix the channel header's actions overlapping the RHS. The inline badge position is harder to solve because the message's content is not predictable since user can use formatting. So for example, what should happen to the badge when a successive reply starts with a title (and to make the matter worse, spans on several lines):

![screen shot 2017-02-26 at 21 20 45](https://cloud.githubusercontent.com/assets/2291025/23343313/88856b74-fc69-11e6-8219-98706203a232.png)

Let me know what you guys come up with. I plan to spare two hours tomorrow between 12:00-14:00 CET to make the appropriate changes (I may have some time later in the day but not sure at this point).

jasonblais commented:
@Zhouzi Thank you for getting the latest fixes in, much appreciated.

I will ask @asaadmahmood to help take a look at this next to help us out here. 

We're almost there! 👍

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0178ff307430fa564.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0178ff307430fa564

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0d0df570d51a92570.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0d0df570d51a92570

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b3388bb2a8b8a9c6.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b3388bb2a8b8a9c6

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-00f28866af717eb8e.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-00f28866af717eb8e

Zhouzi commented:
I'd like to point out an other issue: the "Pinned" badge overlaps the message's content in Safari:

![screen shot 2017-02-28 at 09 29 49](https://cloud.githubusercontent.com/assets/2291025/23397387/9991ab50-fd98-11e6-948a-b1e298fce857.png)

Whereas it's doing fine in Chrome:

![screen shot 2017-02-28 at 09 29 57](https://cloud.githubusercontent.com/assets/2291025/23397403/a30adfe4-fd98-11e6-92f9-6b2181faad66.png)

jasonblais commented:
Thanks @Zhouzi, I've created a ticket for it since it only reproduces on Mac/Safari. https://mattermost.atlassian.net/browse/PLT-5672

Couldn't reproduce on Safari mobile browser

jwilander commented on store/sql_channel_store.go:
```diff
@@ -326,6 +326,32 @@ func (s SqlChannelStore) Get(id string, allowFromCache bool) StoreChannel {
 	return s.get(id, false, allowFromCache)
 }
 
+func (s SqlChannelStore) GetPinnedPosts(channelId string) StoreChannel {
```
This function needs a unit test

Zhouzi commented:
@crspeller @jwilander I addressed the 3 different points:

1. IsPinned is now a boolean (instead of an int)
2. There's now an index on Post.IsPinned
3. I added unit test for sql's GetPinnedPosts

jasonblais commented:
@Zhouzi I tried reaching out to you on https://pre-release.mattermost.com yesterday, I'll post here as well.

First, a huge thanks for contributing this feature! Our team is delighted about the work you've done and we are excited to add it to our product. It's something many of our users have been looking forward to.

However, many of our developers are worried that the `storePost` client function is being modified a lot on this PR which is particularly critical to our product. There is nothing wrong with how you implemented it (no need for changes), but the worry is that it might introduce bugs we don't catch until one or two weeks after the PR is merged.

Given we're only a couple of days away from code complete, it means we might introduce a critical bug that goes undiscovered until very near, or even after, the release is out.

As a measure of precaution, the decision is to hold off on merging this PR for v3.7 but take it early for the next feature release. This let's us
 - test the changes on this PR in https://pre-release.mattermost.com for a couple of weeks before code complete
 - avoid you and me scrambling to fix an unexpected issue right before the release that might be a side effect of modifying the `storePost` function
 - ensure the feature is a success with high quality when we release it

Zhouzi commented:
@jasonblais Thank you very much for the explanation, I appreciate the transparency. It's definitely a reasonable choice and I'm happy we'll have the time to test it properly before an official release.

Just let me know if there's anything I can do to push things forward.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-03336a62845ffd59e.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-03336a62845ffd59e

jasonblais commented:
Thanks @Zhouzi for understanding, it's really appreciated.

For this PR, it's all good to go. I've added a "work in progress" label until I've got confirmation on when we're okay to merge it to `master`. But no further changes are needed (apart from resolving merge conflcts, if any) as it's both PM and dev approved now.

If you'd like, there are these two tickets with some UI improvements/fixes for the feature, one of them is for the Safari edge case you found. If you're interested working on them in a separate PR, that would be a great help:
 - https://mattermost.atlassian.net/browse/PLT-5662
 - https://mattermost.atlassian.net/browse/PLT-5672

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

jasonblais commented:
Got confirmation that we can take this feature in for v3.8

Given PM and dev reviews are complete, applied the `Ready to Merge` label.

@Zhouzi Can you help resolve merge conflicts after which it should be ready to merge?

mattermod commented:
Spinmint test server created at: http://i-09e9113ced90180de.spinmint.com:8065/pr4217

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-09e9113ced90180de

Zhouzi commented:
Hey @jasonblais, sorry for the delay.

The conflicts are now resolved and the branch is up to date with master 🎉 

I'll try to play around with the issues and create the relevant PR if I find anything.

jasonblais commented:
Awesome, thank you!If you'd be able to get a PR for https://mattermost.atlassian.net/browse/PLT-5662 in the next week, that would be awesome :)

grundleborg commented:
Thanks very much for the PR @Zhouzi . This has now been merged into master.

jostyee commented:
Hey,
w/ this PR the mm server fails to start:
`Failed to create index Error 1072: Key column 'IsPinned' doesn't exist in table`
any help?

jwilander commented:
Hi @jostyee, we've noticed the issue and are working on submitting a patch to fix. If you're on pre-release, the thread discussing it is here: https://pre-release.mattermost.com/core/pl/hbwy8d3r5trtmrp9x8x1qw4qoc
